src/driver-hidpp20.c

```diff
@@ -943,44 +943,66 @@ hidpp20drv_update_resolution_dpi(struct ratbag_resolution *resolution,
 	struct ratbag_device *device = profile->device;
 	struct hidpp20drv_data *drv_data = ratbag_get_drv_data(device);
 	struct hidpp20_sensor *sensor;
-	int rc, i;
+	int i;
+	char outstr[23];
 	int dpi = dpi_x; /* dpi_x == dpi_y if we don't have the individual resolution cap */
 
+	if (ratbag_resolution_has_capability(resolution, RATBAG_RESOLUTION_CAP_DISABLE)) {
+		if (resolution->is_disabled && (dpi != 0)) {
+			dpi = 0;
+			dpi_x = 0;
+			dpi_y = 0;
+		} else if (!(resolution->is_disabled) && (dpi == 0)) {
```

Comment:
Not in this case though. This checks if you enable a resolution which had zero dpi before (meaning it was disabled before ratbagd started). Just needed to set a non-zero value.

Reply:
What you are doing here is effectively ignoring the dpi when I call `hidpp20drv_update_resolution_dpi(resolution, 0, 0)` and the resolution is enabled. I am not sure if this is the correct place for that.

Other thing, I missed it in the first review but `!(resolution->is_disabled)` is now irrelevant as we are in the `else` of `resolution->is_disabled`.

Reply:
> I am not sure if this is the correct place for that.

Maybe I got this wrong, but this will happen if you (for example) call do `ratbagctl resolution X enable` with `resolution X` a resolution that previously was disabled and had 0 dpi. Since we don't want any device specific tweaks in the common code (meaning to check when I enable if the resolution is zero dpi), I think it the right place.

Reply:
> Maybe I got this wrong, but this will happen if you (for example) call do ratbagctl resolution X enable with resolution X a resolution that previously was disabled and had 0 dpi. Since we don't want any device specific tweaks in the common code (meaning to check when I enable if the resolution is zero dpi), I think it the right place.

Yes it will. And that's one of the reasons I wanted to have a default sane value for those cases, it will simply things. The resolution should be verified before comitting to the device.

Reply:
> The resolution should be verified before comitting to the device.
But the verification should come from the driver. Basically this depends on the result of the other discussion.

Reply:
> But the verification should come from the driver. Basically this depends on the result of the other discussion.

No? 0dpi is never valid. Also, we export a list of supported dpi values, that should be checked before the command ever reaching the driver.

Reply:
> > But the verification should come from the driver. Basically this depends on the result of the other discussion.
> 
> No? 0dpi is never valid. Also, we export a list of supported dpi values, that should be checked before the command ever reaching the driver.

Well if the device reports and accepts it, it isn't really invalid, but that's not the point.
Here you never _actively_ tell the driver to set zero dpi. You only tell him to enable it, and then the driver validates the dpi. If it was zero before, because it just wan't set aka it was reported like this by the device, then it should set it to some valid value. The user will edit it anyway, so the actual value doesn't really matter. For example a couple of lines below that we check if the sensor accepts the dpi - that's not in the common code.